File: Day1_Q2.java

/*
Given an array of integers and an integer k, determine the number of (i, j) pairs where i < j and the sum of the pair is divisible by k.

Inputs:
An array of integers.

An integer k.

Output:
The number of valid (i, j) pairs.
Example:
arr = [1, 3, 2, 6, 1, 2]

k = 3

Output: 5
 */
public class Day1_Q2 {
    public static int countPairs(int[] arr, int k) {
        int count = 0;
            // Iterate over all pairs (i, j) where i < j
        for (int i = 0; i < arr.length; i++) {
            for (int j = i + 1; j < arr.length; j++) {
                if ((arr[i] + arr[j]) % k == 0) {
                    count++;
                }
            }
        }
        return count;
    }

    public static void main(String[] args) {
        int[] arr = {1, 3, 2, 6, 1, 2};
        int k = 3;
        int result = countPairs(arr, k);
        System.out.println(result);
    }

}

File: DSA/Brijanshi/Day1_Q2.c
#include <stdio.h>

int count_pairs(int arr[], int n, int k) {
    int count[k];
    for (int i = 0; i < k; i++) {
        count[i] = 0;
    }
    for (int i = 0; i < n; i++) {
        count[arr[i] % k]++;
    }
    int pairs = 0;
    for (int i = 0; i <= k / 2; i++) {
        if (i == 0 || i == k / 2) {
            pairs += count[i] * (count[i] - 1) / 2;
        } else {
            pairs += count[i] * count[k - i];
        }
    }
    return pairs;
}

int main() {
    int arr[] = {1, 3, 2, 6, 1, 2};
    int n = sizeof(arr) / sizeof(arr[0]);
    int k = 3;
    int pairs = count_pairs(arr, n, k);
    printf("Number of valid pairs: %d\n", pairs);
    return 0;
}
